File: streampark-flink/streampark-flink-connector/streampark-flink-connector-mongo/src/main/scala/org/apache/streampark/flink/connector/mongo/internal/MongoSourceFunction.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.mongo.internal

import org.apache.streampark.common.enums.ApiType
import org.apache.streampark.common.enums.ApiType.ApiType
import org.apache.streampark.common.util.{Logger, MongoConfig}
import org.apache.streampark.flink.connector.function.RunningFunction
import org.apache.streampark.flink.connector.mongo.function.{MongoQueryFunction, MongoResultFunction}
import org.apache.streampark.flink.util.FlinkUtils

import com.mongodb.MongoClient
import com.mongodb.client.{FindIterable, MongoCollection, MongoCursor}
import org.apache.flink.api.common.state.ListState
import org.apache.flink.api.common.typeinfo.TypeInformation
import org.apache.flink.configuration.Configuration
import org.apache.flink.runtime.state.{CheckpointListener, FunctionInitializationContext, FunctionSnapshotContext}
import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction
import org.apache.flink.streaming.api.functions.source.RichSourceFunction
import org.apache.flink.streaming.api.functions.source.SourceFunction.SourceContext
import org.bson.Document

import java.lang
import java.util.Properties

import org.apache.streampark.common.util.Implicits._
import scala.util.{Success, Try}

class MongoSourceFunction[R: TypeInformation](
    apiType: ApiType,
    prop: Properties = new Properties(),
    collection: String)
  extends RichSourceFunction[R]
  with CheckpointedFunction
  with CheckpointListener
  with Logger {

  @volatile private[this] var running = true
  private[this] var scalaRunningFunc: Unit => Boolean = _
  private[this] var javaRunningFunc: RunningFunction = _

  var client: MongoClient = _
  var mongoCollection: MongoCollection[Document] = _

  private[this] var scalaQueryFunc: (R, MongoCollection[Document]) => FindIterable[Document] = _
  private[this] var scalaResultFunc: MongoCursor[Document] => List[R] = _

  private[this] var javaQueryFunc: MongoQueryFunction[R] = _
  private[this] var javaResultFunc: MongoResultFunction[R] = _

  @transient private var state: ListState[R] = _
  private val OFFSETS_STATE_NAME: String = "mongo-source-query-states"
  private[this] var last: R = _

  // for Scala
  def this(
      collectionName: String,
      prop: Properties,
      scalaQueryFunc: (R, MongoCollection[Document]) => FindIterable[Document],
      scalaResultFunc: MongoCursor[Document] => List[R],
      runningFunc: Unit => Boolean) = {

    this(ApiType.scala, prop, collectionName)
    this.scalaQueryFunc = scalaQueryFunc
    this.scalaResultFunc = scalaResultFunc
    this.scalaRunningFunc = if (runningFunc == null) _ => true else runningFunc
  }

  // for JAVA
  def this(
      collectionName: String,
      prop: Properties,
      queryFunc: MongoQueryFunction[R],
      resultFunc: MongoResultFunction[R],
      runningFunc: RunningFunction) {

    this(ApiType.java, prop, collectionName)
    this.javaQueryFunc = queryFunc
    this.javaResultFunc = resultFunc
    this.javaRunningFunc =
      if (runningFunc != null) runningFunc
      else
        new RunningFunction {
          override def running(): lang.Boolean = true
        }
  }

  override def cancel(): Unit = this.running = false

  override def open(parameters: Configuration): Unit = {
    client = MongoConfig.getClient(prop)
    val db = MongoConfig.getProperty(prop, MongoConfig.database)
    val database = client.getDatabase(db)
    mongoCollection = database.getCollection(collection)
  }

  @throws[Exception]
  override def run(ctx: SourceContext[R]): Unit = {
    while (this.running) {
      apiType match {
        case ApiType.scala =>
          if (scalaRunningFunc()) {
            ctx.getCheckpointLock.synchronized {
              val find = scalaQueryFunc(last, mongoCollection)
              if (find != null) {
                scalaResultFunc(find.iterator).foreach(
                  x => {
                    last = x
                    ctx.collectWithTimestamp(last, System.currentTimeMillis())
                  })
              }
            }
          }
        case ApiType.java =>
          if (javaRunningFunc.running()) {
            ctx.getCheckpointLock.synchronized {
              val find = javaQueryFunc.query(last, mongoCollection)
              if (find != null) {
                javaResultFunc
                  .result(find.iterator)
                  .foreach(
                    x => {
                      last = x
                      ctx.collectWithTimestamp(last, System.currentTimeMillis())
                    })
              }
            }
          }
      }
    }
  }

  override def close(): Unit = {
    client.close()
  }

  override def snapshotState(context: FunctionSnapshotContext): Unit = {
    if (running) {
      state.clear()
      if (last != null) {
        state.add(last)
      }
    } else {
      logError("MongoSource snapshotState called on closed source")
    }
  }

  override def initializeState(context: FunctionInitializationContext): Unit = {
    // restore from checkpoint
    logInfo("MongoSource snapshotState initialize")
    state = FlinkUtils.getUnionListState[R](context, OFFSETS_STATE_NAME)
    Try(state.get.head) match {
      case Success(q) => last = q
      case _ =>
    }
  }

  override def notifyCheckpointComplete(checkpointId: Long): Unit = {
    logInfo(s"MongoSource checkpointComplete: $checkpointId")
  }

}

File: streampark-flink/streampark-flink-connector/streampark-flink-connector-mongo/src/main/java/org/apache/streampark/flink/connector/mongo/internal/MongoSourceFunction.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.flink.connector.mongo.internal;

import org.apache.streampark.common.util.MongoConfig;
import org.apache.streampark.flink.connector.function.RunningFunction;
import org.apache.streampark.flink.connector.mongo.function.MongoQueryFunction;
import org.apache.streampark.flink.connector.mongo.function.MongoResultFunction;
import org.apache.streampark.flink.util.FlinkUtils;

import com.mongodb.MongoClient;
import com.mongodb.client.FindIterable;
import com.mongodb.client.MongoCollection;
import com.mongodb.client.MongoCursor;
import org.apache.flink.api.common.state.ListState;
import org.apache.flink.api.common.typeinfo.TypeInformation;
import org.apache.flink.configuration.Configuration;
import org.apache.flink.runtime.state.CheckpointListener;
import org.apache.flink.runtime.state.FunctionInitializationContext;
import org.apache.flink.runtime.state.FunctionSnapshotContext;
import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
import org.apache.flink.streaming.api.functions.source.SourceFunction;
import org.bson.Document;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.Iterator;
import java.util.List;
import java.util.Properties;

/** MongoDB source function with checkpoint state support. */
public class MongoSourceFunction<R> extends RichSourceFunction<R>
        implements CheckpointedFunction, CheckpointListener {

    private static final Logger LOG = LoggerFactory.getLogger(MongoSourceFunction.class);
    private static final String OFFSETS_STATE_NAME = "mongo-source-query-states";

    private volatile boolean running = true;
    private final Properties prop;
    private final String collection;
    private final MongoQueryFunction<R> queryFunc;
    private final MongoResultFunction<R> resultFunc;
    private final RunningFunction runningFunc;
    private final TypeInformation<R> typeInfo;

    private MongoClient client;
    private MongoCollection<Document> mongoCollection;
    private transient ListState<R> state;
    private R last;

    public MongoSourceFunction(
            String collectionName,
            Properties prop,
            MongoQueryFunction<R> queryFunc,
            MongoResultFunction<R> resultFunc,
            RunningFunction runningFunc,
            TypeInformation<R> typeInfo) {
        this.collection = collectionName;
        this.prop = prop;
        this.queryFunc = queryFunc;
        this.resultFunc = resultFunc;
        this.runningFunc =
                runningFunc != null
                        ? runningFunc
                        : new RunningFunction() {
                            @Override
                            public Boolean running() {
                                return true;
                            }
                        };
        this.typeInfo = typeInfo;
    }

    @Override
    public void open(Configuration parameters) {
        client = MongoConfig.getClient(prop);
        String db = MongoConfig.getProperty(prop, MongoConfig.DATABASE);
        mongoCollection = client.getDatabase(db).getCollection(collection);
    }

    @Override
    public void run(SourceContext<R> ctx) throws Exception {
        while (running) {
            if (runningFunc.running()) {
                synchronized (ctx.getCheckpointLock()) {
                    FindIterable<Document> find = queryFunc.query(last, mongoCollection);
                    if (find != null) {
                        MongoCursor<Document> cursor = find.iterator();
                        List<R> results = new java.util.ArrayList<>();
                        for (R r : resultFunc.result(cursor)) {
                            results.add(r);
                        }
                        for (R x : results) {
                            last = x;
                            ctx.collectWithTimestamp(last, System.currentTimeMillis());
                        }
                    }
                }
            }
        }
    }

    @Override
    public void cancel() {
        running = false;
    }

    @Override
    public void close() {
        if (client != null) {
            client.close();
        }
    }

    @Override
    public void snapshotState(FunctionSnapshotContext context) throws Exception {
        if (running) {
            state.clear();
            if (last != null) {
                state.add(last);
            }
        } else {
            LOG.error("MongoSource snapshotState called on closed source");
        }
    }

    @Override
    public void initializeState(FunctionInitializationContext context) throws Exception {
        LOG.info("MongoSource snapshotState initialize");
        @SuppressWarnings("unchecked")
        TypeInformation<R> stateType =
                typeInfo != null ? typeInfo : (TypeInformation<R>) TypeInformation.of(Object.class);
        state = FlinkUtils.getUnionListState(context, OFFSETS_STATE_NAME, stateType);
        Iterator<R> it = state.get().iterator();
        if (it.hasNext()) {
            last = it.next();
        }
    }

    @Override
    public void notifyCheckpointComplete(long checkpointId) {
        LOG.info("MongoSource checkpointComplete: {}", checkpointId);
    }
}
